### PR TITLE
Upgrading the project to ESNext

### PR DESCRIPTION
**Pull request by NicoVogel**
I upgraded the project to allow for the use of modern JS. While this PR is big, I tried to keep it as small as possible.

# What did I add?

1. Support for both `CommonJS` and `ESM`
2. Allow coding in `ESNext` and transpile with `babel`
3. Convert the code from `CommonJS` to `ESM` (with minimal changes)
4. VSCode debugging support
5. Nearly removed all `eslint` disable lines
6. Added `prettier` to ensure a uniformly formatted code base (and aligned `eslint` with `prettier`)

# Why did I make these changes?

- It is future proof
- It enforces one coding format
- Simplifies debugging for VSCode developer

# Each change in detail

## 1. Support for both `CommonJS` and `ESM` 

With this change its possible to import oas-tools as `CommonJS` or `ESM`, depending on the use case. I followed this [article](https://2ality.com/2019/10/hybrid-npm-packages.html) to set it up. Another aspect which I considered is to get rid of the `.npmignore` file. The article [For the love of god, don’t use .npmignore](https://medium.com/@jdxcode/for-the-love-of-god-dont-use-npmignore-f93c08909d8d) does a good job at explaining why you should use the `files` property instead of `.npmignore`.

The next step was to decide, which target versions should be used for `CommonJS` and `ESM` respectively. For this I used the [node releases overview page](https://nodejs.org/en/about/releases/) as reference. This resulted in the following target versions:

  - `CommonJS` targets Node `10.23` _(could also be lower, but this is the lowest still supported version)_
  - `ESM` targets Node `14.15` _(could also be 12.x, but modules are hidden behind an experimental flag)_

## 2. Allow coding in `ESNext` and transpile with `babel`

In order to allow for this hybrid approach a transpiler is needed, which takes care of the version compatibilities. Therefore I added babel to the project. It allows to transpile the `src` directory into `CommonJS` (transpiled into `common`) and `ESM` (transpiled into `lib`). While babel allows to write `ESNext` code, eslint also needs to be aware of the fact that the code is in fact following the `ESNext` standard. 

## 3. Convert the code from `CommonJS` to `ESM` (with minimal changes)

After enabling the tooling, the next step was to convert the existing code into `ES Modules` format. This resulted in quite some code changes which are mostly minor (The logic changes are described below). It's mostly changing form `require` to `import` and from `module.exports` to `export`. Sometimes there was the need to restructure the export parts, because `export` is not as flexible as `module.exports`.

## 4. VSCode debugging support

This was simply adding a babel build config for live building the tests when executing them and adding the `.vscode` folder to provide the needed configuration.

## 5. Nearly removed all `eslint` disable lines

While updating the code to newer standards is great. A general goal should always be to follow the linter rules defined for a project. Therefore I removed all but one `//eslint-disable-next-line`. Why not the last one? It's in `oas-router.js:336` and the only way of fixing it is to get rid of the `prefer-rest-params` rule. I tried keeping all the rules if possible and only change the code accordingly. But in this case, I am not sure why arguments are used. I believe it's not needed, because all function parameters are passed anyways....

I left it to come back later and properly resolve it.

## 6. Added `prettier` to ensure a uniformly formatted code base (and aligned `eslint` with `prettier`)

The final step was to ensure a consistent code structure/look/feel. The best way of archiving it is by using `prettier` in combination with `eslint`. The setup only required removing some `eslint` rules which interfered with `prettier`. The best thing is that `prettier` per default uses the `.editor.config` file as a base configuration and therefore no `.prettierrc` file is needed.

I then added the script `format` to allow formatting the entire project. Because of this change, only double quotes are left in the project and many indentations have changed. **Therefore, I suggest reviewing with ignore whitespace!!**

# Changes which are not directly related to the goal of this PR

- The test server was updated to get the tests passing again. Only a bare minimum was done, because my next PR will address the testing. It's nice to have tests, but putting it all in one file is hard to understand and maintain.
- The validator got some bug fixes along the way, because it was either needed to get tests passing again or convenient to fix while fixing `//eslint-disable-next-line`.
  - Parameters which are not required are new validated _(see line 149)_ [#250 fixed]
  - The nullable keyword is taken into consideration when validating _(see line 155)_ [#234 could be fixed, but not sure] 
  - Headers are considered case insensitive _(see line 117)_

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,3 +1,3 @@
-FROM node:6.9
+FROM node:10.23.0-alpine
```
Comment (tuunit) [inline Dockerfile:1]: @raffrearaUS We actually don't now if this Dockerfile is used for anything?

@NicoVogel You updated the node version to 12 in travis but here it is 10. Any reason for that?

**Review by tuunit — APPROVED** [review]

**Review by tuunit — COMMENTED** [review]
@raffrearaUS @NicoVogel I would suggest a minor or even major version upgrade of the project after this PR

File: Dockerfile (hunk 1 above)
Comment (raffrearaUS) [reply Dockerfile:1]: The Dockerfile is part of the Node template we used for the project, but it is not used for anything and can be safely removed.

**Comment by raffrearaUS** [comment]
Sure, if there are no more changes to be done I will merge the PR and then release version 2.2.0.

File: Dockerfile (hunk 1 above)
Comment (NicoVogel) [reply Dockerfile:1]: Then I would remove it as part of this PR.

**Comment by NicoVogel** [comment]
I want to validate whether a native ESM import works, before you merge it. Currently the tests use the CommonJS import, so if you would merge now, no existing project will break with this change. But I did not yet have time to properly validate if the new ESM import works as intended. So, I will inform you when I am done with that.

**Pull request merged**